dmq-node/src/DMQ/Protocol/SigSubmission/Validate.hs

```diff
@@ -0,0 +1,217 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE MultiWayIf         #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE PatternSynonyms    #-}
+{-# LANGUAGE TupleSections      #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE TypeOperators      #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- | Encapsulates signature validation utilities leveraged by the mempool writer
+--
+module DMQ.Protocol.SigSubmission.Validate where
+
+import Control.Monad
+import Control.Concurrent.Class.MonadSTM.Strict
+import Control.Exception (Exception)
+import Control.Monad.Class.MonadTime.SI
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Except
+import Control.Monad.Trans.Except.Extra
+import Data.Aeson
+import Data.ByteString (ByteString)
+import Data.ByteString.Lazy qualified as LBS
+import Data.Map.Strict qualified as Map
+import Data.Maybe (fromJust, isNothing)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Typeable
+import Data.Word
+
+import Cardano.Crypto.DSIGN.Class (ContextDSIGN)
+import Cardano.Crypto.DSIGN.Class qualified as DSIGN
+import Cardano.Crypto.KES.Class (KESAlgorithm (..))
+import Cardano.KESAgent.KES.Crypto as KES
+import Cardano.KESAgent.KES.OCert (OCert (..), OCertSignable, validateOCert)
+import Cardano.Ledger.BaseTypes.NonZero
+import Cardano.Ledger.Hashes
+
+import DMQ.Diffusion.NodeKernel (PoolValidationCtx (..))
+import DMQ.Protocol.SigSubmission.Type
+import Ouroboros.Consensus.Shelley.Ledger.Query
+import Ouroboros.Network.TxSubmission.Mempool.Simple
+import Ouroboros.Network.Util.ShowProxy
+
+
+-- | The type of non-fatal failures reported by the mempool writer
+-- for invalid messages
+--
+data instance TxValidationFail (Sig crypto) =
+    SigInvalid SigValidationError
+  | SigDuplicate
+  | SigExpired
+  | SigResultOther Text
+  deriving (Eq, Show)
+
+instance (Typeable crypto) => ShowProxy (TxValidationFail (Sig crypto))
+
+instance (Typeable crypto) => Exception (TxValidationFail (Sig crypto))
+
+instance ToJSON (TxValidationFail (Sig crypto)) where
+  toJSON SigDuplicate = String "duplicate"
+  toJSON SigExpired   = String "expired"
+  toJSON (SigInvalid e) = object
+    [ "type" .= String "invalid"
+    , "reason" .= show e
+    ]
+  toJSON (SigResultOther txt) = object
+    [ "type" .= String "other"
+    , "reason" .= txt
+      ]
+
+
+data SigValidationError =
+    InvalidKESSignature KESPeriod KESPeriod String
+  | InvalidSignatureOCERT
+      !Word64    -- OCert counter
+      !KESPeriod -- OCert KES period
+      !String    -- DSIGN error message
+  | InvalidOCertCounter
+      Word64 -- last seen
+      Word64 -- received
```

Comment:
Just for consistency make these fields strict as well, or remove the strictness annotation from the other ones.  I bet they are evaluated strictly anyway :grin:.